### PR TITLE
ROS Docker

### PR DESCRIPTION
**Pull request by utkarshjp7**
Wrote a docker script for running sawyer experiments. The script has been tested in simulation. The instructions for using the docker script can be found here `gym-sawyer/sawyer/ros/docker/README.md`.

### CODE REVIEW

File: sawyer/ros/docker/docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#start gazebo sawyer
+cd $ROS_WS && ./intera.sh sim "export PYTHONPATH=/opt/ros/kinetic/lib/python2.7/dist-packages:$ROS_WS/devel/lib/python3/dist-packages && roslaunch sawyer_gazebo sawyer_learning.launch &"
+sleep 5
+
+#start sawyer moveit
+cd $ROS_WS && ./intera.sh sim "roslaunch sawyer_moveit_config sawyer_moveit.launch electric_gripper:=true &"
+sleep 5
+
+#change default python version to 3.6
+ln -fs /usr/bin/python3.6 /usr/bin/python
```
Comment (ryanjulian) [inline sawyer/ros/docker/docker-entrypoint.sh]: Again this is incredibly hacky. Please try to find another way. Mangling symlinks is a terrible way to switch Python versions.

File: sawyer/ros/docker/run.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
```
Comment (ryanjulian) [inline sawyer/ros/docker/run.sh]: How does this differ from the entrypoint?

File: sawyer/ros/docker/run_gpu.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
+
```
Comment (ryanjulian) [inline sawyer/ros/docker/run_gpu.sh]: Isn't there a docker project which does this stuff for you? docker-opengl?

File: sawyer/ros/docker/Dockerfile.gpu [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+
+# Add code stub last
+RUN rm -rf /root/code/gym-sawyer
+ 
+COPY . /root/code/gym-sawyer
+
+ENV PYTHONPATH=$PYTHONPATH:/root/code/gym-sawyer
+
+RUN /root/code/gym-sawyer/sawyer/ros/docker/setup_ros_workspace.sh
```
Comment (ryanjulian) [inline sawyer/ros/docker/Dockerfile.gpu]: Do not call giant setup scripts inside dockerfiles. The whole point of a dockerfile is that it's a setup script. All the commands go in here.

File: sawyer/ros/docker/Dockerfile.gpu [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+ joblib \
+ protobuf \
+ jsonmerge \
+ tensorboard \
+ tensorflow \
+"]
+
+# Prepare Intera SDK installation
+# Install dependencies
+RUN apt update && DEBIAN_FRONTEND=noninteractive apt install -y \
```
Comment (ryanjulian) [inline sawyer/ros/docker/Dockerfile.gpu]: Generally every Dockerfile should have a *single* layer for APT dependencies, a *single* layer for pip dependencies, etc.

File: sawyer/ros/docker/run_gpu.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+if [ -z ${NVIDIA_DRIVER+x} ]; then
+	echo "Error: Could not determine NVIDIA driver version number. Please specify your driver version number manually in $0." 1>&2 ;
+	exit ;
+else
+	echo "Linking to NVIDIA driver version $NVIDIA_DRIVER..." ;
+fi
+
+DOCKER_VISUAL_NVIDIA="-v /tmp/.X11-unix:/tmp/.X11-unix --device /dev/nvidia0 --device /dev/nvidiactl"
+
+docker run \
```
Comment (eric-heiden) [inline sawyer/ros/docker/run_gpu.sh]: If this is supposed to run on a GPU, why are you not using nvidia-docker? Did you verify that this run script will actually enable you to use the GPU from within the docker?

File: sawyer/ros/docker/Dockerfile.gpu [hunk 7]
```diff
@@ -0,0 +1,233 @@
+ARG DOCKER_FROM=ubuntu:16.04
+
+FROM ubuntu:16.04 as glvnd
+MAINTAINER Eric Heiden <me@eric-heiden.com>
```
Comment (eric-heiden) [inline sawyer/ros/docker/Dockerfile.gpu]: I know a lot has been copied from https://github.com/eric-heiden/docker-robot-learning but I'm not the maintainer of this docker file.

**Comment by ryanjulian** [comment]
Docker already provides a built-in composition command--the `FROM` directive. Anything which doesn't fit into `FROM` (or `ENV` or `ARG`) should just be a different Dockerfile.

If we start to get an explosion of Dockerfiles I could imagine supporting `#include` macros using m4 and a Makefile (like [this](http://bobbynorton.com/posts/includes-in-dockerfiles-with-m4-and-make/)), but I don't feel we're at that point yet.

Sometimes the right decision is just to copy.

## More than you wanted to know
Docker omits general templating support as an intentional design decision to preserve two core principles of the system:
1. You can recreate any container at any time using only its Dockerfile
1. Containers are always stateless (between invocations)

Most of the time people experience trouble and frustration using Docker, it's because they are trying to violate one of these two principles (the rest is usually their bad CLI interface).

Dockerfiles are a linear, declarative specification of system dependencies for stateless containers, not a weird scripting language for configuring stateful VMs. If you add general templating, the system is no longer declarative, and I can't create a container using only the Dockerfile.

The way Docker is designed, the only way to "template" is to organize your description of the system from least-to-most specific. You can put the platform-specific parts in upstream Dockerfiles and include them using the `FROM` directive.

tl;dr: Please don't template your Dockerfiles or use giant shell scripts inside them. Most of the time, it leads to an unmaintainble mess. It'd much rather maintain 2 copies of a Dockerfile than have to

File: sawyer/ros/docker/run_gpu.sh [hunk 8]
```diff
@@ -3,6 +3,9 @@
 USER_UID=$(id -u)
 USER_GID=$(id -g)
 
+SAWYER_HOSTNAME="021707CP00056.local"
```
Comment (ryanjulian) [inline sawyer/ros/docker/run_gpu.sh]: where do these IDs com from? I think you should retrieve them from command-line args or environment variables.

check out https://argbash.io/ for an easy way to add argument parsing to your script.

File: sawyer/ros/docker/Dockerfile.cpu [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+# MuJoCo
+RUN mkdir /root/.mujoco && \
+  wget https://www.roboti.us/download/mjpro150_linux.zip -O mujoco.zip && \
+  unzip mujoco.zip -d $HOME/.mujoco && \
+  rm mujoco.zip
+ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/root/.mujoco/mjpro150/bin
+
+#Install Python 3.6
+RUN apt-add-repository ppa:deadsnakes/ppa
```
Comment (gonzaiva) [inline sawyer/ros/docker/Dockerfile.cpu]: I'm getting the following error when apt-add-repository gets called:
```
Step 16/31 : RUN apt-add-repository ppa:deadsnakes/ppa
 ---> Running in 4480e4180b99
/bin/sh: 1: apt-add-repository: not found
The command '/bin/sh -c apt-add-repository ppa:deadsnakes/ppa' returned a non-zero code: 127
```
I think we have to install software-properties-common first as shown [here](https://stackoverflow.com/a/32487026).

File: sawyer/ros/docker/Dockerfile.cpu [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+RUN rosdep init && rosdep update
+# install bootstrap tools
+RUN apt update && DEBIAN_FRONTEND=noninteractive apt install -y \
+  python-rosinstall \
+  python-wstool \
+  python-yaml && \
+  apt-get clean && \
+  rm -rf /var/lib/apt/lists/*
+
+RUN ["/bin/bash", "-c", "python3.6 -m pip install \
```
Comment (gonzaiva) [inline sawyer/ros/docker/Dockerfile.cpu]: I'm also getting an error in this command:
```
Step 20/27 : RUN ["/bin/bash", "-c", "python3.6 -m pip install  catkin_pkg  rospkg  defusedxml  opencv-python  empy  gym  joblib  protobuf  jsonmerge  tensorboard  tensorflow  "]
 ---> Running in 625b0d803ef0
/usr/bin/python3.6: No module named pip
The command '/bin/bash -c python3.6 -m pip install  catkin_pkg  rospkg  defusedxml  opencv-python  empy  gym  joblib  protobuf  jsonmerge  tensorboard  tensorflow  ' returned a non-zero code: 1
```
By removing `python3.6 -m` and only calling `pip install` the problem is solved.

File: sawyer/ros/docker/Dockerfile.cpu (hunk 9 above)
Comment (utkarshjp7) [reply sawyer/ros/docker/Dockerfile.cpu]: @gonzaiva  I am currently in the process of dividing this one huge docker file into smaller, more maintainable docker containers. A lot of these long bash commands will be gone after this change.
